File: operator/src/resources.rs
//! Kubernetes resource builders for SSD-KV cluster components.

use std::collections::BTreeMap;

use k8s_openapi::api::apps::v1::{StatefulSet, StatefulSetSpec};
use k8s_openapi::api::core::v1::{
    ConfigMap, Container, ContainerPort, EnvVar, PersistentVolumeClaim,
    PodSpec, PodTemplateSpec, Service, ServicePort, ServiceSpec, VolumeMount,
};
use k8s_openapi::apimachinery::pkg::api::resource::Quantity;
use k8s_openapi::apimachinery::pkg::apis::meta::v1::{LabelSelector, ObjectMeta};
use kube::ResourceExt;

use crate::controller::SsdkvCluster;
use crate::shard_manager::ShardMap;

fn cluster_labels(cluster: &SsdkvCluster) -> BTreeMap<String, String> {
    let mut labels = BTreeMap::new();
    labels.insert("app".to_string(), "ssd-kv".to_string());
    labels.insert(
        "ssdkv.io/cluster".to_string(),
        cluster.name_any(),
    );
    labels
}

/// Builds the headless Service for StatefulSet pod DNS.
pub fn build_headless_service(cluster: &SsdkvCluster) -> Service {
    let name = format!("{}-headless", cluster.name_any());
    let namespace = cluster.namespace().unwrap_or_else(|| "default".to_string());
    let labels = cluster_labels(cluster);

    Service {
        metadata: ObjectMeta {
            name: Some(name),
            namespace: Some(namespace),
            labels: Some(labels.clone()),
            ..Default::default()
        },
        spec: Some(ServiceSpec {
            cluster_ip: Some("None".to_string()),
            selector: Some(labels),
            ports: Some(vec![
                ServicePort {
                    name: Some("redis".to_string()),
                    port: cluster.spec.redis_port,
                    ..Default::default()
                },
                ServicePort {
                    name: Some("cluster".to_string()),
                    port: cluster.spec.cluster_port,
                    ..Default::default()
                },
            ]),
            ..Default::default()
        }),
        ..Default::default()
    }
}

/// Builds the client-facing Service (load balanced across all nodes).
pub fn build_client_service(cluster: &SsdkvCluster) -> Service {
    let name = cluster.name_any();
    let namespace = cluster.namespace().unwrap_or_else(|| "default".to_string());
    let labels = cluster_labels(cluster);

    Service {
        metadata: ObjectMeta {
            name: Some(name),
            namespace: Some(namespace),
            labels: Some(labels.clone()),
            ..Default::default()
        },
        spec: Some(ServiceSpec {
            selector: Some(labels),
            ports: Some(vec![ServicePort {
                name: Some("redis".to_string()),
                port: cluster.spec.redis_port,
                ..Default::default()
            }]),
            ..Default::default()
        }),
        ..Default::default()
    }
}

/// Builds a read-only client Service for replica read traffic.
pub fn build_readonly_service(cluster: &SsdkvCluster) -> Service {
    let name = format!("{}-readonly", cluster.name_any());
    let namespace = cluster.namespace().unwrap_or_else(|| "default".to_string());
    let mut labels = cluster_labels(cluster);
    labels.insert("ssdkv.io/role".to_string(), "replica".to_string());

    Service {
        metadata: ObjectMeta {
            name: Some(name),
            namespace: Some(namespace),
            labels: Some(labels.clone()),
            ..Default::default()
        },
        spec: Some(ServiceSpec {
            selector: Some(labels),
            ports: Some(vec![ServicePort {
                name: Some("redis".to_string()),
                port: cluster.spec.redis_port,
                ..Default::default()
            }]),
            ..Default::default()
        }),
        ..Default::default()
    }
}

/// Builds the topology ConfigMap containing shard-to-node mapping.
pub fn build_topology_configmap(cluster: &SsdkvCluster, shard_map: &ShardMap) -> ConfigMap {
    let name = format!("{}-topology", cluster.name_any());
    let namespace = cluster.namespace().unwrap_or_else(|| "default".to_string());
    let labels = cluster_labels(cluster);

    let mut data = BTreeMap::new();
    data.insert("total_nodes".to_string(), cluster.spec.replicas.to_string());
    data.insert(
        "replication_factor".to_string(),
        cluster.spec.replication_factor.to_string(),
    );
    data.insert(
        "shard_map.json".to_string(),
        serde_json::to_string(&shard_map.assignments).unwrap_or_default(),
    );

    // Generate peer list
    let headless_svc = format!("{}-headless", cluster.name_any());
    let peers: Vec<String> = (0..cluster.spec.replicas)
        .map(|i| {
            format!(
                "{}-{}.{}.{}.svc.cluster.local:{}",
                cluster.name_any(),
                i,
                headless_svc,
                namespace,
                cluster.spec.cluster_port,
            )
        })
        .collect();
    data.insert("peers".to_string(), peers.join(","));

    ConfigMap {
        metadata: ObjectMeta {
            name: Some(name),
            namespace: Some(namespace),
            labels: Some(labels),
            ..Default::default()
        },
        data: Some(data),
        ..Default::default()
    }
}

/// Builds the StatefulSet for SSD-KV nodes.
pub fn build_statefulset(cluster: &SsdkvCluster) -> StatefulSet {
    let name = cluster.name_any();
    let namespace = cluster.namespace().unwrap_or_else(|| "default".to_string());
    let labels = cluster_labels(cluster);
    let headless_svc = format!("{}-headless", name);

    // Build peer list for --cluster-peers
    let peers: Vec<String> = (0..cluster.spec.replicas)
        .map(|i| {
            format!(
                "{}-{}.{}.{}.svc.cluster.local:{}",
                name, i, headless_svc, namespace, cluster.spec.cluster_port
            )
        })
        .collect();
    let peers_str = peers.join(",");

    let replica_read_flag = if cluster.spec.replica_read {
        " --replica-read"
    } else {
        ""
    };

    // Build resource requirements if specified
    let container_resources = cluster.spec.resources.as_ref().map(|res| {
        let mut reqs = BTreeMap::new();
        let mut lims = BTreeMap::new();
        if let Some(ref r) = res.requests {
            if let Some(ref cpu) = r.cpu {
                reqs.insert("cpu".to_string(), Quantity(cpu.clone()));
            }
            if let Some(ref mem) = r.memory {
                reqs.insert("memory".to_string(), Quantity(mem.clone()));
            }
        }
        if let Some(ref l) = res.limits {
            if let Some(ref cpu) = l.cpu {
                lims.insert("cpu".to_string(), Quantity(cpu.clone()));
            }
            if let Some(ref mem) = l.memory {
                lims.insert("memory".to_string(), Quantity(mem.clone()));
            }
        }
        k8s_openapi::api::core::v1::ResourceRequirements {
            requests: if reqs.is_empty() { None } else { Some(reqs) },
            limits: if lims.is_empty() { None } else { Some(lims) },
            ..Default::default()
        }
    });

    let container = Container {
        name: "ssd-kv".to_string(),
        image: Some(cluster.spec.image.clone()),
        image_pull_policy: Some("IfNotPresent".to_string()),
        resources: container_resources,
        readiness_probe: Some(k8s_openapi::api::core::v1::Probe {
            tcp_socket: Some(k8s_openapi::api::core::v1::TCPSocketAction {
                port: k8s_openapi::apimachinery::pkg::util::intstr::IntOrString::Int(cluster.spec.redis_port),
                ..Default::default()
            }),
            initial_delay_seconds: Some(5),
            period_seconds: Some(10),
            ..Default::default()
        }),
        liveness_probe: Some(k8s_openapi::api::core::v1::Probe {
            tcp_socket: Some(k8s_openapi::api::core::v1::TCPSocketAction {
                port: k8s_openapi::apimachinery::pkg::util::intstr::IntOrString::Int(cluster.spec.redis_port),
                ..Default::default()
            }),
            initial_delay_seconds: Some(15),
            period_seconds: Some(20),
            ..Default::default()
        }),
        ports: Some(vec![
            ContainerPort {
                name: Some("redis".to_string()),
                container_port: cluster.spec.redis_port,
                ..Default::default()
            },
            ContainerPort {
                name: Some("cluster".to_string()),
                container_port: cluster.spec.cluster_port,
                ..Default::default()
            },
        ]),
        env: Some(vec![
            EnvVar {
                name: "POD_NAME".to_string(),
                value_from: Some(k8s_openapi::api::core::v1::EnvVarSource {
                    field_ref: Some(k8s_openapi::api::core::v1::ObjectFieldSelector {
                        field_path: "metadata.name".to_string(),
                        ..Default::default()
                    }),
                    ..Default::default()
                }),
                ..Default::default()
            },
        ]),
        command: Some(vec!["/bin/sh".to_string()]),
        args: Some(vec![
            "-c".to_string(),
            format!(
                concat!(
                    "NODE_ID=$(echo $POD_NAME | rev | cut -d'-' -f1 | rev) && ",
                    "exec ssd-kv ",
                    "--bind 0.0.0.0:{} ",
                    "--data-dir /data ",
                    "--cluster-mode ",
                    "--node-id $NODE_ID ",
                    "--total-nodes {} ",
                    "--replication-factor {} ",
                    "--cluster-port {} ",
                    "--cluster-peers {}",
                    "{}"
                ),
                cluster.spec.redis_port,
                cluster.spec.replicas,
                cluster.spec.replication_factor,
                cluster.spec.cluster_port,
                peers_str,
                replica_read_flag,
            ),
        ]),
        volume_mounts: Some(vec![VolumeMount {
            name: "data".to_string(),
            mount_path: "/data".to_string(),
            ..Default::default()
        }]),
        ..Default::default()
    };

    // Build PVC template
    let mut pvc_resources = BTreeMap::new();
    pvc_resources.insert(
        "storage".to_string(),
        Quantity(cluster.spec.storage.size.clone()),
    );

    let pvc_template = PersistentVolumeClaim {
        metadata: ObjectMeta {
            name: Some("data".to_string()),
            ..Default::default()
        },
        spec: Some(k8s_openapi::api::core::v1::PersistentVolumeClaimSpec {
            access_modes: Some(vec!["ReadWriteOnce".to_string()]),
            resources: Some(k8s_openapi::api::core::v1::VolumeResourceRequirements {
                requests: Some(pvc_resources),
                ..Default::default()
            }),
            storage_class_name: cluster.spec.storage.storage_class.clone(),
            ..Default::default()
        }),
        ..Default::default()
    };

    StatefulSet {
        metadata: ObjectMeta {
            name: Some(name.clone()),
            namespace: Some(namespace),
            labels: Some(labels.clone()),
            ..Default::default()
        },
        spec: Some(StatefulSetSpec {
            replicas: Some(cluster.spec.replicas),
            selector: LabelSelector {
                match_labels: Some(labels.clone()),
                ..Default::default()
            },
            service_name: headless_svc,
            template: PodTemplateSpec {
                metadata: Some(ObjectMeta {
                    labels: Some(labels),
                    ..Default::default()
                }),
                spec: Some(PodSpec {
                    containers: vec![container],
                    ..Default::default()
                }),
            },
            volume_claim_templates: Some(vec![pvc_template]),
            ..Default::default()
        }),
        ..Default::default()
    }
}

File: operator/resources.go
package main

import (
	"fmt"
	"strings"

	appsv1 "k8s.io/api/apps/v1"
	corev1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/api/resource"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/util/intstr"
)

func headlessServiceName(cluster *SsdkvCluster) string {
	return cluster.Name + "-headless"
}

func buildHeadlessService(cluster *SsdkvCluster) *corev1.Service {
	redisPort := cluster.Spec.GetRedisPort()
	clusterPort := cluster.Spec.GetClusterPort()

	return &corev1.Service{
		ObjectMeta: metav1.ObjectMeta{
			Name:      headlessServiceName(cluster),
			Namespace: cluster.Namespace,
			Labels:    labels(cluster),
		},
		Spec: corev1.ServiceSpec{
			ClusterIP: "None",
			Selector:  labels(cluster),
			Ports: []corev1.ServicePort{
				{Name: "redis", Port: redisPort, TargetPort: intstr.FromInt32(redisPort), Protocol: corev1.ProtocolTCP},
				{Name: "cluster", Port: clusterPort, TargetPort: intstr.FromInt32(clusterPort), Protocol: corev1.ProtocolTCP},
			},
			PublishNotReadyAddresses: true,
		},
	}
}

func buildClientService(cluster *SsdkvCluster) *corev1.Service {
	redisPort := cluster.Spec.GetRedisPort()

	return &corev1.Service{
		ObjectMeta: metav1.ObjectMeta{
			Name:      cluster.Name,
			Namespace: cluster.Namespace,
			Labels:    labels(cluster),
		},
		Spec: corev1.ServiceSpec{
			Type:     corev1.ServiceTypeClusterIP,
			Selector: labels(cluster),
			Ports: []corev1.ServicePort{
				{Name: "redis", Port: redisPort, TargetPort: intstr.FromInt32(redisPort), Protocol: corev1.ProtocolTCP},
			},
		},
	}
}

func buildStatefulSet(cluster *SsdkvCluster) (*appsv1.StatefulSet, error) {
	replicas := cluster.Spec.GetReplicas()
	redisPort := cluster.Spec.GetRedisPort()
	clusterPort := cluster.Spec.GetClusterPort()
	rf := cluster.Spec.GetReplicationFactor()

	// Build --cluster-peers DNS list
	headlessSvc := headlessServiceName(cluster)
	var peers []string
	for i := int32(0); i < replicas; i++ {
		// Pod DNS: <name>-<ordinal>.<headless-svc>.<namespace>.svc.cluster.local:<cluster-port>
		peers = append(peers, fmt.Sprintf("%s-%d.%s.%s.svc.cluster.local:%d",
			cluster.Name, i, headlessSvc, cluster.Namespace, clusterPort))
	}
	peerList := strings.Join(peers, ",")

	// Container command: extract ordinal from hostname
	command := []string{"/bin/sh", "-c", fmt.Sprintf(
		`ORDINAL=$(echo $HOSTNAME | rev | cut -d'-' -f1 | rev) && `+
			`exec /usr/local/bin/ssd-kv `+
			`--bind 0.0.0.0:%d `+
			`--data-dir /data `+
			`--cluster-mode `+
			`--node-id $ORDINAL `+
			`--total-nodes %d `+
			`--cluster-port %d `+
			`--cluster-peers %s `+
			`--replication-factor %d`,
		redisPort, replicas, clusterPort, peerList, rf)}

	if cluster.Spec.ReplicaRead {
		command[2] += " --replica-read"
	}

	container := corev1.Container{
		Name:    "ssdkv",
		Image:   cluster.Spec.Image,
		Command: command,
		Ports: []corev1.ContainerPort{
			{Name: "redis", ContainerPort: redisPort, Protocol: corev1.ProtocolTCP},
			{Name: "cluster", ContainerPort: clusterPort, Protocol: corev1.ProtocolTCP},
		},
		ReadinessProbe: &corev1.Probe{
			ProbeHandler: corev1.ProbeHandler{
				TCPSocket: &corev1.TCPSocketAction{Port: intstr.FromInt32(redisPort)},
			},
			InitialDelaySeconds: 5,
			PeriodSeconds:       5,
		},
		LivenessProbe: &corev1.Probe{
			ProbeHandler: corev1.ProbeHandler{
				TCPSocket: &corev1.TCPSocketAction{Port: intstr.FromInt32(redisPort)},
			},
			InitialDelaySeconds: 15,
			PeriodSeconds:       10,
		},
	}

	// Resources are required so every pod is Guaranteed-QoS and eligible for
	// the kubelet's static CPU manager.
	if cluster.Spec.Resources == nil {
		return nil, fmt.Errorf("spec.resources is required (cpu and memory)")
	}
	rr, err := buildResourceRequirements(cluster.Spec.Resources)
	if err != nil {
		return nil, err
	}
	container.Resources = rr

	// Volume mount
	container.VolumeMounts = []corev1.VolumeMount{
		{Name: "data", MountPath: "/data"},
	}

	ss := &appsv1.StatefulSet{
		ObjectMeta: metav1.ObjectMeta{
			Name:      cluster.Name,
			Namespace: cluster.Namespace,
			Labels:    labels(cluster),
		},
		Spec: appsv1.StatefulSetSpec{
			ServiceName: headlessServiceName(cluster),
			Replicas:    &replicas,
			Selector:    &metav1.LabelSelector{MatchLabels: labels(cluster)},
			Template: corev1.PodTemplateSpec{
				ObjectMeta: metav1.ObjectMeta{Labels: labels(cluster)},
				Spec: corev1.PodSpec{
					Containers: []corev1.Container{container},
				},
			},
		},
	}

	// PVC template or emptyDir
	if cluster.Spec.Storage != nil && cluster.Spec.Storage.Size != "" {
		storageQty := resource.MustParse(cluster.Spec.Storage.Size)
		pvc := corev1.PersistentVolumeClaim{
			ObjectMeta: metav1.ObjectMeta{Name: "data"},
			Spec: corev1.PersistentVolumeClaimSpec{
				AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
				Resources: corev1.VolumeResourceRequirements{
					Requests: corev1.ResourceList{corev1.ResourceStorage: storageQty},
				},
			},
		}
		if cluster.Spec.Storage.StorageClassName != "" {
			pvc.Spec.StorageClassName = &cluster.Spec.Storage.StorageClassName
		}
		ss.Spec.VolumeClaimTemplates = []corev1.PersistentVolumeClaim{pvc}
	} else {
		ss.Spec.Template.Spec.Volumes = []corev1.Volume{
			{Name: "data", VolumeSource: corev1.VolumeSource{EmptyDir: &corev1.EmptyDirVolumeSource{}}},
		}
	}

	return ss, nil
}

// buildResourceRequirements returns a Guaranteed-QoS ResourceRequirements:
// requests == limits for both CPU and memory. CPU must be a positive integer
// (whole cores) so the kubelet's static CPU manager can assign exclusive cores.
func buildResourceRequirements(r *ResourceSpec) (corev1.ResourceRequirements, error) {
	cpuQty, err := resource.ParseQuantity(r.CPU)
	if err != nil {
		return corev1.ResourceRequirements{}, fmt.Errorf("invalid cpu %q: %w", r.CPU, err)
	}
	if cpuQty.MilliValue()%1000 != 0 || cpuQty.Sign() <= 0 {
		return corev1.ResourceRequirements{}, fmt.Errorf(
			"cpu must be a positive whole number of cores (got %q); fractional CPUs disable static CPU pinning",
			r.CPU,
		)
	}
	memQty, err := resource.ParseQuantity(r.Memory)
	if err != nil {
		return corev1.ResourceRequirements{}, fmt.Errorf("invalid memory %q: %w", r.Memory, err)
	}

	list := corev1.ResourceList{
		corev1.ResourceCPU:    cpuQty,
		corev1.ResourceMemory: memQty,
	}
	return corev1.ResourceRequirements{Requests: list, Limits: list}, nil
}

func labels(cluster *SsdkvCluster) map[string]string {
	return map[string]string{
		"app.kubernetes.io/name":       "ssdkv",
		"app.kubernetes.io/instance":   cluster.Name,
		"app.kubernetes.io/managed-by": "ssdkv-operator",
	}
}
